File: Medium Problem's/p-53 Maximum Subarray/Solution.c
#include <stdio.h>
#include <limits.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))

// LeetCode formate Solution
int maxSubArray(int *nums, int numsSize)
{
    int maxSum = INT_MIN;
    int curSum = 0;
    for (int i = 0; i < numsSize; i++)
    {
        curSum += nums[i];
        maxSum = MAX(curSum, maxSum);
        if (curSum < 0)
        {
            curSum = 0;
        }
    }

    return maxSum;
}

int main(int argc, char const *argv[])
{

    int nums[] = {5, 4, -1, 7, 8};
    int size = sizeof(nums) / sizeof(nums[0]);

    int result = maxSubArray(nums, size);
    printf("the max sum is %d", result);

    return 0;
}

File: Medium Problem's/p-53 Maximum Subarray/Solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */

//Satyam-R4j-LeetCode Problem-53 (medium) Solution
//Solution in JavaScript -> Two-Pointer Approach
var maxSubArray = function(nums) {
    let max = -Infinity
    let sum = 0
    for(let i = 0; i < nums.length; i++)
    {
        sum = sum + nums[i]
        max = Math.max(max , sum)
        if(sum < 0)
        {
            sum = 0
        }
    }
    return max
};

let nums = [-2,1,-3,4,-1,2,1,-5,4]
console.log(maxSubArray(nums));
